spring-petclinic-api-gateway/src/test/java/org/springframework/samples/petclinic/api/application/VisitsServiceClientTest.java

```diff
@@ -0,0 +1,86 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.client.circuitbreaker.EnableCircuitBreaker;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.EnableAspectJAutoProxy;
+import org.springframework.http.MediaType;
+import org.springframework.samples.petclinic.api.dto.VisitDetails;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.springframework.test.web.client.MockRestServiceServer;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo;
+import static org.springframework.test.web.client.response.MockRestResponseCreators.withServerError;
+import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;
+
+@ExtendWith(SpringExtension.class)
+@EnableCircuitBreaker
+@EnableAspectJAutoProxy
+@ContextConfiguration(classes = {VisitsServiceClient.class})
+class VisitsServiceClientTest {
+
+    @Autowired
+    private VisitsServiceClient visitsServiceClient;
+
+    @Autowired
+    private RestTemplate restTemplate;
+
+    private MockRestServiceServer mockServer;
+
+    @Configuration
+    static class Config {
+        @Bean
+        public RestTemplate restTemplate() {
+            return new RestTemplate();
+        }
+    }
+
+
+    @BeforeEach
+    public void setUp() {
+        mockServer = MockRestServiceServer.createServer(restTemplate);
+    }
+
+
+    @Test
+    public void getVisitsForPets_withAvailableVisitsService() {
+        mockServer.expect(requestTo("http://visits-service/pets/visits?petId=1"))
+            .andRespond(withSuccess("{\"items\":[{\"id\":5,\"date\":\"2018-11-15\",\"description\":\"test visit\",\"petId\":1}]}", MediaType.APPLICATION_JSON));
+
+        Map<Integer, List<VisitDetails>> visits = visitsServiceClient.getVisitsForPets(Collections.singletonList(1));
+
+        assertNotNull(visits);
+        assertEquals(1, visits.size());
+        assertNotNull(visits.get(1));
+        assertEquals(1, visits.get(1).size());
+        assertEquals("test visit", visits.get(1).get(0).getDescription());
+    }
+
+    /**
+     * Test Hystrix fallback method
+     */
+    @Test
+    public void getVisitsForPets_withServerError() {
+
+        mockServer.expect(requestTo("http://visits-service/pets/visits?petId=1"))
+            .andRespond(withServerError());
+
+        Map<Integer, List<VisitDetails>> visits = visitsServiceClient.getVisitsForPets(Collections.singletonList(1));
+
+        assertNotNull(visits);
+        assertEquals(0, visits.size());
```

Comment:
I think first assertion is redundant

Reply:
If map is `null`, then `assertEquals(0, visits.size());` signals error and fails a build. Less assertions - less noise in tests.

Reply:
That's a point of view. I prefer to fail the test as soon as possible. But if done the change.